numpy/lib/arraysetops.py

```diff
@@ -82,7 +82,12 @@ def ediff1d(ary, to_end=None, to_begin=None):
     ary = np.asanyarray(ary).ravel()
 
     # enforce that the dtype of `ary` is used for the output
-    dtype_req = ary.dtype
+    # The difference of np.datetime64 will yield an np.timedelta64 as the
+    # required dtype for to_begin and to_end
+    if np.issubdtype(ary.dtype, np.datetime64):
+        dtype_req = ary.astype(np.timedelta64).dtype
+    else:
+        dtype_req = ary.dtype
```

Comment:
The `ary.astype()` would make a full copy just to get the correct output dtype, which I think needs to be improved upon.

I would love to reorganize the code to use the correct result dtype directly.  Unfortunately, there is no good way to get that (yet!) besides doing a (partial) calculation: `(ary[:1] - ary[:1]).dtype`.

We could use that as a hack rather than the `astype()`, since right now I cannot think of another way that doesn't use the string representatio.

I would then like to have a comment like:
```
# TODO: Once exposed, may use `np.subtract.resolve_descriptors()`
#       (or similar) to find the result dtype and generalize this.
```

Reply:
Maybe we can just leave that `if ary.dtype.kind == "M"` that you had for now?  I suspect this is always fine, but it is not really ideal (yet) and it will be easy to relax further in the future.

Reply:
Sure sounds good, I've explicitly checked for a `np.datetime64` in https://github.com/numpy/numpy/pull/21788/commits/61bc077f8d612b1a416793ded3692e23fe061f0a